File: vultisig-analytics/ingestors/arkham_ingestor.py
"""
Arkham DEX Aggregator Ingestor
Fetches transfer data from Arkham API and ingests into database.
Uses actual fee amounts from Arkham as ground truth.
"""

import os
import logging
import requests
import psycopg2
from datetime import datetime
from typing import List, Dict, Optional
from .protocol_identifier import ProtocolIdentifier

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Configuration
ARKHAM_API_KEY = os.getenv('ARKHAM_API_KEY')
ARKHAM_API_BASE = 'https://api.arkhamintelligence.com'
INTEGRATOR_ADDRESS = '0xA4a4f610e89488EB4ECc6c63069f241a54485269'
DATABASE_URL = os.getenv('DATABASE_URL')

# Chain name normalization
CHAIN_MAPPING = {
    'ethereum': 'Ethereum',
    'bsc': 'BSC',
    'binance-smart-chain': 'BSC',
    'polygon': 'Polygon',
    'polygon-pos': 'Polygon',
    'arbitrum_one': 'Arbitrum',
    'arbitrum-one': 'Arbitrum',
    'optimism': 'Optimism',
    'base': 'Base',
    'avalanche': 'Avalanche',
    'blast': 'Blast',
}

class ArkhamIngestor:
    """Ingests DEX aggregator revenue data from Arkham API."""
    
    def __init__(self):
        if not ARKHAM_API_KEY:
            raise ValueError("ARKHAM_API_KEY environment variable not set")

        if not DATABASE_URL:
            raise ValueError("DATABASE_URL environment variable not set")

        self.api_key = ARKHAM_API_KEY
        self.database_url = DATABASE_URL
        self.db = None
        self.protocol_identifier = None

    def _get_connection(self):
        """Get or refresh database connection"""
        try:
            # Test if connection is alive
            if self.db and not self.db.closed:
                cursor = self.db.cursor()
                cursor.execute("SELECT 1")
                cursor.close()
                return self.db
        except (psycopg2.OperationalError, psycopg2.InterfaceError):
            logger.warning("Database connection lost, reconnecting...")
            if self.db:
                try:
                    self.db.close()
                except:
                    pass

        # Create new connection
        self.db = psycopg2.connect(self.database_url)
        self.protocol_identifier = ProtocolIdentifier(self.db)
        logger.info("Database connection established")
        return self.db
        
    def fetch_all_transfers(self) -> List[Dict]:
        """
        Fetch transfers from Arkham API.
        Only fetches new transfers by checking latest timestamp in database.
        """
        url = f'{ARKHAM_API_BASE}/transfers'
        headers = {
            'Accept': 'application/json',
            'API-Key': self.api_key
        }
        
        # Get latest timestamp from database to avoid re-fetching
        try:
            db = self._get_connection()
            cursor = db.cursor()
            cursor.execute("SELECT MAX(timestamp) FROM dex_aggregator_revenue WHERE fee_data_source = 'arkham'")
            result = cursor.fetchone()
            latest_timestamp = result[0] if result and result[0] else None
            if latest_timestamp:
                logger.info(f"Latest Arkham timestamp in DB: {latest_timestamp}")
            cursor.close()
        except Exception as e:
            logger.warning(f"Could not fetch latest timestamp: {e}")
            latest_timestamp = None
        
        all_transfers = []
        offset = 0
        limit = 1000
        
        logger.info(f"Fetching transfers for integrator {INTEGRATOR_ADDRESS}")
        
        while True:
            params = {
                'base': INTEGRATOR_ADDRESS,
                'limit': limit,
                'offset': offset
            }
            
            try:
                response = requests.get(url, params=params, headers=headers, timeout=30)
                response.raise_for_status()
                
                data = response.json()
                transfers = data.get('transfers', [])
                
                if not transfers:
                    break
                
                # Filter out transfers we already have
                new_transfers = []
                for transfer in transfers:
                    timestamp_str = transfer.get('blockTimestamp')
                    if timestamp_str and latest_timestamp:
                        try:
                            transfer_time = datetime.fromisoformat(timestamp_str.replace('Z', '+00:00'))
                            if transfer_time <= latest_timestamp:
                                # We've reached transfers we already have, stop fetching
                                logger.info(f"Reached existing transfers at {transfer_time}, stopping fetch")
                                return all_transfers
                        except:
                            pass
                    new_transfers.append(transfer)
                
                all_transfers.extend(new_transfers)
                logger.info(f"Fetched {len(new_transfers)} new transfers (offset {offset})")
                
                offset += len(transfers)
                
                # If we got fewer than limit, we're done
                if len(transfers) < limit:
                    break
                    
            except requests.exceptions.RequestException as e:
                logger.error(f"Error fetching transfers: {e}")
                break
        
        logger.info(f"Total new transfers fetched: {len(all_transfers)}")
        return all_transfers
    
    def normalize_chain(self, chain: str) -> str:
        """Normalize chain name to match our database format."""
        if not chain:
            return 'Unknown'
        return CHAIN_MAPPING.get(chain.lower(), chain.capitalize())
    
    def extract_address(self, addr_obj) -> str:
        """
        Extract address string from Arkham's address object.
        Arkham returns addresses as objects with nested structure.
        """
        if not addr_obj:
            return ''
        if isinstance(addr_obj, str):
            return addr_obj
        if isinstance(addr_obj, dict):
            return addr_obj.get('address', '')
        return ''
    
    def identify_protocol_from_arkham_entity(self, addr_obj) -> Optional[str]:
        """
        Extract protocol name from Arkham's arkhamEntity metadata.
        This is more reliable than address matching.
        """
        if not isinstance(addr_obj, dict):
            return None
        
        entity = addr_obj.get('arkhamEntity')
        if not entity or not isinstance(entity, dict):
            return None
        
        entity_id = entity.get('id', '').lower()
        entity_name = entity.get('name', '').lower()
        
        # Map Arkham entity names to our protocol names
        if '1inch' in entity_id or '1inch' in entity_name:
            return '1inch'
        elif 'paraswap' in entity_id or 'paraswap' in entity_name:
            return 'paraswap'
        elif 'cow' in entity_id or 'cowswap' in entity_name:
            return 'cowswap'
        elif 'matcha' in entity_id or '0x' in entity_name:
            return 'matcha'
        
        return None
    
    def insert_transfer(self, transfer: Dict):
        """
        Insert a single transfer into the database.
        Stores ALL available fields from Arkham API response.

        Args:
            transfer: Transfer data from Arkham API
        """
        try:
            tx_hash = transfer.get('transactionHash')
            if not tx_hash:
                logger.warning(f"Transfer missing transaction hash: {transfer}")
                return

            # Extract addresses from objects
            from_address_obj = transfer.get('fromAddress', {})
            from_address = self.extract_address(from_address_obj)
            to_address = self.extract_address(transfer.get('toAddress', {}))

            # Extract chain and normalize
            chain = self.normalize_chain(transfer.get('chain', ''))

            # Try to identify protocol from Arkham entity first (most reliable)
            protocol = self.identify_protocol_from_arkham_entity(from_address_obj)

            # If not found, use our protocol identifier
            if not protocol:
                protocol = self.protocol_identifier.identify_protocol(tx_hash, from_address, chain)

            # Extract fee data (what Arkham provides directly)
            actual_fee_usd = float(transfer.get('historicalUSD', 0))
            fee_token_symbol = transfer.get('tokenSymbol', '')
            fee_token_address = transfer.get('tokenAddress', '')
            fee_amount_raw = str(transfer.get('unitValue', ''))

            # Parse timestamp
            timestamp = transfer.get('blockTimestamp')
            if timestamp:
                timestamp = datetime.fromisoformat(timestamp.replace('Z', '+00:00'))
            else:
                timestamp = datetime.now()

            block_number = transfer.get('blockNumber')

            # NEW: Extract ALL available token/swap data from Arkham response
            # Note: Arkham may not provide swap details, only fee transfers
            # Token in/out data would need to be enriched from blockchain RPC
            token_in_address = None
            token_out_address = None
            token_in_symbol = None
            token_out_symbol = None
            amount_in = None
            amount_out = None
            swap_volume_usd = None
            volume_data_source = None

            # Some Arkham transfers may include transaction details
            # Check if this is a native token transfer (ETH, BNB, etc.)
            if fee_token_address == 'NATIVE' or not fee_token_address:
                token_in_address = 'NATIVE'
                # For native transfers, use the chain's native token symbol
                chain_native_tokens = {
                    'Ethereum': 'ETH',
                    'BSC': 'BNB',
                    'Polygon': 'MATIC',
                    'Arbitrum': 'ETH',
                    'Optimism': 'ETH',
                    'Base': 'ETH',
                    'Avalanche': 'AVAX',
                    'Blast': 'ETH'
                }
                token_in_symbol = chain_native_tokens.get(chain, '')

            # If Arkham provides unit value and we know the price, calculate amount
            unit_value = transfer.get('unitValue')
            if unit_value and actual_fee_usd > 0:
                try:
                    amount_in = float(unit_value)
                    # swap_volume_usd would be the total swap amount (not fee)
                    # We don't have this from Arkham directly
                except:
                    pass

            # Insert into database (or update if exists)
            db = self._get_connection()
            cursor = db.cursor()
            cursor.execute("""
                INSERT INTO dex_aggregator_revenue (
                    tx_hash, chain, protocol, timestamp,
                    actual_fee_usd, fee_token_symbol, fee_token_address,
                    fee_amount_raw,
                    swap_volume_usd, token_in_symbol, token_in_address,
                    token_out_symbol, token_out_address,
                    amount_in, amount_out,
                    block_number, from_address, to_address,
                    fee_data_source, volume_data_source
                ) VALUES (
                    %s, %s, %s, %s,
                    %s, %s, %s,
                    %s,
                    %s, %s, %s,
                    %s, %s,
                    %s, %s,
                    %s, %s, %s,
                    %s, %s
               )
                ON CONFLICT (tx_hash) DO UPDATE SET
                    actual_fee_usd = EXCLUDED.actual_fee_usd,
                    protocol = EXCLUDED.protocol,
                    fee_token_symbol = EXCLUDED.fee_token_symbol,
                    token_in_symbol = EXCLUDED.token_in_symbol,
                    token_in_address = EXCLUDED.token_in_address,
                    amount_in = EXCLUDED.amount_in,
                    fee_data_source = EXCLUDED.fee_data_source,
                    volume_data_source = EXCLUDED.volume_data_source,
                    updated_at = NOW()
            """, (
                tx_hash, chain, protocol, timestamp,
                actual_fee_usd, fee_token_symbol, fee_token_address,
                fee_amount_raw,
                swap_volume_usd, token_in_symbol, token_in_address,
                token_out_symbol, token_out_address,
                amount_in, amount_out,
                block_number,
                from_address,
                to_address,
                'arkham',
                volume_data_source
            ))

            logger.debug(f"Inserted/Updated {tx_hash} - {protocol} on {chain}: ${actual_fee_usd:.2f}")

        except Exception as e:
            logger.error(f"Error inserting transfer {transfer.get('transactionHash')}: {e}")
            logger.error(f"Transfer data: {transfer}")
            if self.db:
                self.db.rollback()
    
    def ingest(self):
        """Main ingestion process."""
        try:
            logger.info("Starting Arkham ingestion")

            # Ensure connection is established
            db = self._get_connection()

            # Fetch all transfers
            transfers = self.fetch_all_transfers()

            if not transfers:
                logger.warning("No transfers fetched")
                return

            # Insert each transfer
            for i, transfer in enumerate(transfers, 1):
                self.insert_transfer(transfer)

                # Commit every 100 records and refresh connection
                if i % 100 == 0:
                    db.commit()
                    logger.info(f"Committed {i}/{len(transfers)} transfers")
                    # Refresh connection to prevent staleness
                    db = self._get_connection()

            # Final commit
            db.commit()
            logger.info(f"Completed ingestion of {len(transfers)} transfers")

            # Print stats
            if self.protocol_identifier:
                stats = self.protocol_identifier.get_protocol_stats()
                logger.info("Protocol breakdown:")
                for protocol, count in stats.items():
                    logger.info(f"  {protocol}: {count} transactions")

        except Exception as e:
            logger.error(f"Error during ingestion: {e}")
            if self.db:
                self.db.rollback()
        finally:
            if self.db:
                self.db.close()

def main():
    """Entry point for Arkham ingestor."""
    ingestor = ArkhamIngestor()
    ingestor.ingest()

if __name__ == '__main__':
    main()

File: vultisig-analytics/ingestors/arkham_ingestor.js
/**
 * Arkham DEX Aggregator Ingestor (Node.js version)
 * Fetches transfer data from Arkham API and ingests into database.
 * Uses actual fee amounts from Arkham as ground truth.
 */

const axios = require('axios');
const { Pool } = require('pg');
require('dotenv').config();

// Configuration
const ARKHAM_API_KEY = process.env.ARKHAM_API_KEY;
const ARKHAM_API_BASE = 'https://api.arkhamintelligence.com';
const INTEGRATOR_ADDRESS = '0xA4a4f610e89488EB4ECc6c63069f241a54485269';
const DATABASE_URL = process.env.DATABASE_URL;

if (!ARKHAM_API_KEY) {
    console.error('❌ ARKHAM_API_KEY environment variable not set');
    process.exit(1);
}

if (!DATABASE_URL) {
    console.error('❌ DATABASE_URL environment variable not set');
    process.exit(1);
}

const pool = new Pool({
    connectionString: DATABASE_URL,
});

// Chain name normalization mapping
const CHAIN_MAPPING = {
    'ethereum': 'Ethereum',
    'bsc': 'BSC',
    'binance-smart-chain': 'BSC',
    'polygon': 'Polygon',
    'polygon-pos': 'Polygon',
    'arbitrum_one': 'Arbitrum',
    'arbitrum-one': 'Arbitrum',
    'optimism': 'Optimism',
    'base': 'Base',
    'avalanche': 'Avalanche',
    'blast': 'Blast',
};

// Known DEX aggregator router addresses (lowercase)
const KNOWN_ROUTERS = {
    '1inch': [
        '0x1111111254eeb25477b68fb85ed929f73a960582',  // 1inch v5 Router
        '0x111111125421ca6dc452d289314280a0f8842a65',  // 1inch v6 Router (Fusion)
        '0x11111112542d85b3ef69ae05771c2dccff4faa26',  // 1inch v4 Router
    ],
    'paraswap': [
        '0xdef171fe48cf0115b1d80b88dc8eab59176fee57',  // Paraswap Augustus v5
        '0x216b4b4ba9f3e719726886d34a177484278bfcae',  // Paraswap Augustus v6
    ],
    'cowswap': [
        '0x9008d19f58aabd9ed0d60971565aa8510560ab41',  // CoWSwap Settlement
    ],
    'matcha': [
        '0xdef1c0ded9bec7f1a1670819833240f027b25eff',  // 0x Exchange Proxy (Matcha)
    ],
    '0x': [
        '0xdef1c0ded9bec7f1a1670819833240f027b25eff',  // 0x Exchange Proxy
    ],
    'thorchain': [
        '0xd37bbe5744d730a1d98d8dc97c42f0ca46ad7146',  // THORChain Router v4.1.1
    ]
};

// Flatten router map for quick lookup
const ROUTER_TO_PROTOCOL = {};
Object.entries(KNOWN_ROUTERS).forEach(([protocol, addresses]) => {
    addresses.forEach(addr => {
        ROUTER_TO_PROTOCOL[addr.toLowerCase()] = protocol;
    });
});

/**
 * Normalize chain name
 */
function normalizeChain(chain) {
    if (!chain) return 'Unknown';
    return CHAIN_MAPPING[chain.toLowerCase()] || chain.charAt(0).toUpperCase() + chain.slice(1);
}

/**
 * Extract address string from Arkham's address object
 */
function extractAddress(addrObj) {
    if (!addrObj) return '';
    if (typeof addrObj === 'string') return addrObj;
    if (typeof addrObj === 'object') return addrObj.address || '';
    return '';
}

/**
 * Identify protocol from Arkham entity metadata
 */
function identifyProtocolFromEntity(addrObj) {
    if (!addrObj || typeof addrObj !== 'object') return null;

    const entity = addrObj.arkhamEntity;
    if (!entity || typeof entity !== 'object') return null;

    const id = (entity.id || '').toLowerCase();
    const name = (entity.name || '').toLowerCase();

    if (id.includes('1inch') || name.includes('1inch')) return '1inch';
    if (id.includes('paraswap') || name.includes('paraswap')) return 'paraswap';
    if (id.includes('cow') || name.includes('cowswap')) return 'cowswap';
    if (id.includes('matcha') || name.includes('0x')) return 'matcha';

    return null;
}

/**
 * Identify protocol by router address
 */
function identifyProtocolByAddress(address) {
    if (!address) return null;
    return ROUTER_TO_PROTOCOL[address.toLowerCase()] || null;
}

/**
 * Check if transaction exists in 1inch database
 */
async function identifyBy1inchDb(client, txHash) {
    try {
        const res = await client.query(
            "SELECT 1 FROM swaps WHERE source = '1inch' AND LOWER(tx_hash) = LOWER($1) LIMIT 1",
            [txHash]
        );
        return res.rowCount > 0 ? '1inch' : null;
    } catch (err) {
        console.error(`Error checking 1inch DB for ${txHash}:`, err.message);
        return null;
    }
}

const VolumeExtractor = require('./volume_extractor');
const volumeExtractor = new VolumeExtractor();

/**
 * Main ingestion function
 */
async function ingest() {
    const client = await pool.connect();

    try {
        console.log('=== STARTING ARKHAM INGESTION (Node.js) ===');
        console.log(`Integrator: ${INTEGRATOR_ADDRESS}`);

        // Fetch transfers
        let allTransfers = [];
        let offset = 0;
        const limit = 1000;
        let hasMore = true;

        while (hasMore) {
            console.log(`Fetching batch at offset ${offset}...`);

            try {
                const response = await axios.get(`${ARKHAM_API_BASE}/transfers`, {
                    params: {
                        base: INTEGRATOR_ADDRESS,
                        limit,
                        offset
                    },
                    headers: {
                        'Accept': 'application/json',
                        'API-Key': ARKHAM_API_KEY
                    },
                    timeout: 30000
                });

                const transfers = response.data.transfers || [];
                console.log(`  ✅ Got ${transfers.length} transfers`);

                if (transfers.length === 0) {
                    hasMore = false;
                } else {
                    allTransfers.push(...transfers);
                    offset += transfers.length;
                    if (transfers.length < limit) hasMore = false;
                }

                // Rate limiting
                if (hasMore) await new Promise(r => setTimeout(r, 1000));

            } catch (error) {
                console.error('  ❌ API Error:', error.message);
                hasMore = false;
            }
        }

        console.log(`\nTotal transfers fetched: ${allTransfers.length}`);

        // Process and insert
        let processedCount = 0;
        let errorCount = 0;

        // await client.query('BEGIN'); // Removed to prevent single error from aborting all

        for (const transfer of allTransfers) {
            try {
                const txHash = transfer.transactionHash;
                if (!txHash) continue;

                // Extract addresses
                const fromAddressObj = transfer.fromAddress;
                const fromAddress = extractAddress(fromAddressObj);
                const toAddress = extractAddress(transfer.toAddress);

                // FILTER: Only count INCOMING transfers (Revenue)
                if (toAddress.toLowerCase() !== INTEGRATOR_ADDRESS.toLowerCase()) {
                    // console.log(`Skipping outgoing transfer ${txHash}`);
                    continue;
                }

                // Identify protocol
                let protocol = identifyProtocolFromEntity(fromAddressObj);

                if (!protocol) {
                    protocol = identifyProtocolByAddress(fromAddress);
                }

                if (!protocol) {
                    protocol = await identifyBy1inchDb(client, txHash);
                }

                if (!protocol) {
                    protocol = 'other';
                }

                // Normalize chain
                const chain = normalizeChain(transfer.chain);

                // Extract Volume (New Step)
                let volumeData = null;
                // Only try to extract volume for supported protocols/chains to save RPC calls
                if (protocol === '1inch' || protocol === 'paraswap' || protocol === 'cowswap' || protocol === 'other') {
                    try {
                        volumeData = await volumeExtractor.getVolume(txHash, chain);
                    } catch (err) {
                        console.warn(`Failed to extract volume for ${txHash}: ${err.message}`);
                    }
                }

                // Extract fee data
                const actualFeeUsd = parseFloat(transfer.historicalUSD || 0);
                const feeTokenSymbol = transfer.tokenSymbol || '';
                const feeTokenAddress = transfer.tokenAddress || '';
                const feeAmountRaw = String(transfer.unitValue || '');

                // Timestamp
                const timestamp = transfer.blockTimestamp ? new Date(transfer.blockTimestamp) : new Date();

                // Insert
                await client.query(`
          INSERT INTO dex_aggregator_revenue (
            tx_hash, chain, protocol, timestamp,
            actual_fee_usd, fee_token_symbol, fee_token_address,
            fee_amount_raw, block_number, from_address, to_address,
            fee_data_source,
            amount_in, token_in_address, volume_data_source, token_in_decimals
          ) VALUES (
            $1, $2, $3, $4,
            $5, $6, $7,
            $8, $9, $10, $11,
            'arkham',
            $12, $13, $14, $15
          )
          ON CONFLICT (tx_hash) DO UPDATE SET
            actual_fee_usd = EXCLUDED.actual_fee_usd,
            protocol = EXCLUDED.protocol,
            fee_token_symbol = EXCLUDED.fee_token_symbol,
            fee_data_source = EXCLUDED.fee_data_source,
            amount_in = EXCLUDED.amount_in,
            token_in_address = EXCLUDED.token_in_address,
            volume_data_source = EXCLUDED.volume_data_source,
            token_in_decimals = EXCLUDED.token_in_decimals,
            updated_at = NOW()
        `, [
                    txHash,
                    chain,
                    protocol,
                    timestamp,
                    actualFeeUsd,
                    feeTokenSymbol,
                    feeTokenAddress,
                    feeAmountRaw,
                    transfer.blockNumber || null,
                    fromAddress,
                    toAddress,
                    volumeData ? volumeData.amount : null,
                    volumeData ? volumeData.token : null,
                    volumeData ? 'blockchain_rpc' : null,
                    volumeData ? volumeData.decimals : null
                ]);

                processedCount++;
                if (processedCount % 100 === 0) {
                    console.log(`Processed ${processedCount}/${allTransfers.length}`);
                }

            } catch (err) {
                console.error(`Error processing tx ${transfer.transactionHash}:`, err.message);
                errorCount++;
            }
        }

        await client.query('COMMIT');
        console.log(`\n✅ Ingestion complete!`);
        console.log(`Processed: ${processedCount}`);
        console.log(`Errors: ${errorCount}`);

        // Show stats
        const stats = await client.query(`
      SELECT protocol, COUNT(*) as count, SUM(actual_fee_usd) as total_fees
      FROM dex_aggregator_revenue
      GROUP BY protocol
      ORDER BY total_fees DESC
    `);

        console.log('\nProtocol Breakdown:');
        console.table(stats.rows);

    } catch (err) {
        await client.query('ROLLBACK');
        console.error('❌ Fatal error:', err);
    } finally {
        client.release();
        await pool.end();
    }
}

ingest();
